File: py/year2022/day06.py
with open(f"input/year2022/day06.txt", "r") as f:
    line = f.read()

def get_header(packet, n):
    for i, chars in enumerate(zip(*[packet[i:] for i in range(n)])):
        if len(set(chars)) == n:
            return i + n

print(get_header(line, 4))
print(get_header(line, 14))

File: rust/src/bin/year2022/day06.rs
use std::fs;

pub fn run() {
    let mut text = fs::read_to_string("../input/year2022/day06.txt").unwrap();
    text = text.split("\r\n").take(1).collect::<String>();

    let part1 = get_packet_marker_idx(&text, 4);
    let part2 = get_packet_marker_idx(&text, 14);
    println!("{}", part1);
    println!("{}", part2);
}

fn get_packet_marker_idx(text: &String, size: usize) -> i32 {
    for i in 0..(text.len() - size) {
        let chars: u32 = (0..size)
            .map(|j| 1 << text.as_bytes()[i + j] - "a".as_bytes()[0])
            .sum();
        if u32::count_ones(chars) == size as u32 {
            return (i + size) as i32;
        }
    }
    -1
}
